cpp/src/arrow/array/array_nested.cc

```diff
@@ -177,6 +177,8 @@ Result<std::shared_ptr<Array>> FlattenListArray(const ListArrayT& list_array,
   // Final attempt to avoid invoking Concatenate().
   if (non_null_fragments.size() == 1) {
     return non_null_fragments[0];
+  } else if (non_null_fragments.size() == 0) {
+    return MakeEmptyArray(value_array->type(), memory_pool);
```

Comment:
Might also be worth it to add and `} else if (list_array.null_count() == list_array->length()) {` to the `if (list_array.null_count() == 0)` above and avoid the calculation of fragments altogether when we are lucky and `null_count` is already calculated for us.

Reply:
You can force the null calculation in the tests by calling `GetNullCount()`.

Reply:
Done

Reply:
But I think you need to keep both checks. `null_count()` can return `kUnknownNullCount` here. :sweat:

Reply:
Thanks!